File: frontend/src/services/transcription.ts
/**
 * Transcription Service - API calls for transcription operations
 */
import { getClient, Body } from "@tauri-apps/api/http";
import { readBinaryFile } from "@tauri-apps/api/fs";
import api from "./api";
import type {
  JobCreateResponse,
  JobResponse,
  TranscriptionRequest,
  ModelsResponse,
} from "@/types/api";

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || "http://localhost:8000";

/**
 * Upload and transcribe an audio file using Tauri's HTTP client
 * This properly handles multipart form data in Tauri
 */
export async function transcribeFile(
  file: File,
  options: TranscriptionRequest = {}
): Promise<JobCreateResponse> {
  const client = await getClient();
  
  // Read file content as Uint8Array
  const fileBuffer = await file.arrayBuffer();
  const fileData = new Uint8Array(fileBuffer);
  
  // Build form data using Tauri's Body.form with proper file part
  // According to Tauri docs, file can be a path string OR an array buffer
  // The fileName is REQUIRED for Tauri's multipart to work correctly
  const formData = new FormData();
  formData.append("file", new Blob([fileData], { type: file.type || "audio/mpeg" }), file.name);
  
  // Add optional parameters as form fields
  if (options.language) {
    formData.append("language", options.language);
  }
  formData.append("model_size", options.model_size || "base");
  formData.append("enable_diarization", String(options.enable_diarization ?? false));
  if (options.num_speakers) {
    formData.append("num_speakers", String(options.num_speakers));
  }

  const response = await client.post<JobCreateResponse | { error: { code: string; message: string } }>(
    `${API_BASE_URL}/api/v1/transcribe/file`,
    Body.form(formData),
    {
      headers: {
        "Content-Type": "multipart/form-data",
      },
    }
  );

  // Check for error response
  if (!response.ok) {
    const errorData = response.data as { error?: { code: string; message: string } };
    if (errorData?.error) {
      throw new Error(`${errorData.error.code}: ${errorData.error.message}`);
    }
    throw new Error(`Request failed with status ${response.status}`);
  }

  return response.data as JobCreateResponse;
}

/**
 * Alternative: Upload file from path (for Tauri file dialog)
 * This reads the file directly from disk using the path
 */
export async function transcribeFileFromPath(
  filePath: string,
  fileName: string,
  mimeType: string,
  options: TranscriptionRequest = {}
): Promise<JobCreateResponse> {
  const client = await getClient();
  
  // Read file content from disk
  const fileData = await readBinaryFile(filePath);
  
  // Build FormData with the file
  const formData = new FormData();
  formData.append("file", new Blob([fileData.buffer as ArrayBuffer], { type: mimeType }), fileName);
  
  if (options.language) {
    formData.append("language", options.language);
  }
  formData.append("model_size", options.model_size || "base");
  formData.append("enable_diarization", String(options.enable_diarization ?? false));
  if (options.num_speakers) {
    formData.append("num_speakers", String(options.num_speakers));
  }

  const response = await client.post<JobCreateResponse | { error: { code: string; message: string } }>(
    `${API_BASE_URL}/api/v1/transcribe/file`,
    Body.form(formData),
    {
      headers: {
        "Content-Type": "multipart/form-data",
      },
    }
  );

  if (!response.ok) {
    const errorData = response.data as { error?: { code: string; message: string } };
    if (errorData?.error) {
      throw new Error(`${errorData.error.code}: ${errorData.error.message}`);
    }
    throw new Error(`Request failed with status ${response.status}`);
  }

  return response.data as JobCreateResponse;
}

/**
 * Get job status and result
 */
export async function getJobStatus(jobId: string): Promise<JobResponse> {
  const response = await api.get<JobResponse>(`/transcribe/job/${jobId}`);
  return response.data;
}

/**
 * Poll job status until completion or failure
 */
export async function pollJobStatus(
  jobId: string,
  onProgress?: (job: JobResponse) => void,
  pollInterval: number = 2000
): Promise<JobResponse> {
  return new Promise((resolve, reject) => {
    const poll = async () => {
      try {
        const job = await getJobStatus(jobId);

        // Call progress callback if provided
        if (onProgress) {
          onProgress(job);
        }

        // Check if job is complete
        if (job.status === "completed") {
          resolve(job);
        } else if (job.status === "failed") {
          reject(new Error(job.error || "Job failed"));
        } else {
          // Continue polling
          setTimeout(poll, pollInterval);
        }
      } catch (error) {
        reject(error);
      }
    };

    // Start polling
    poll();
  });
}

/**
 * Get available models and languages
 */
export async function getAvailableModels(): Promise<ModelsResponse> {
  const response = await api.get<ModelsResponse>("/models");
  return response.data;
}

/**
 * Check API health
 */
export async function checkHealth(): Promise<boolean> {
  try {
    await api.get("/health");
    return true;
  } catch {
    return false;
  }
}

File: backend/app/api/v1/transcription.py
"""Transcription API endpoints."""
from typing import Optional

from fastapi import APIRouter, File, Form, UploadFile

from app.core.config import settings
from app.core.errors import FileValidationError
from app.core.logging import get_logger
from app.models.transcription import (
    HealthResponse,
    JobCreateResponse,
    JobResponse,
    JobStatus,
    ModelSize,
    ModelsResponse,
    TranscriptionRequest,
)
from app.services.job_manager import get_job_manager
from app.services.transcription import get_transcription_service
from app.utils.audio import AudioProcessor
from app.utils.file_handler import FileHandler

logger = get_logger(__name__)

router = APIRouter()


@router.post("/transcribe/file", response_model=JobCreateResponse)
async def transcribe_file(
    file: UploadFile = File(..., description="Audio file to transcribe"),
    language: Optional[str] = Form(None, description="Source language code"),
    model_size: str = Form("base", description="Whisper model size"),
    enable_diarization: bool = Form(False, description="Enable speaker diarization"),
    num_speakers: Optional[int] = Form(None, description="Number of speakers"),
) -> JobCreateResponse:
    """Upload and transcribe an audio file.

    Args:
        file: Audio file upload
        language: Optional language code
        model_size: Whisper model size
        enable_diarization: Enable speaker diarization
        num_speakers: Number of speakers for diarization

    Returns:
        Job creation response with job ID

    Raises:
        FileValidationError: If file validation fails
    """
    # Log detailed file info for debugging
    logger.info(f"Received transcription request for file: {file.filename}")
    logger.info(f"File content_type: {file.content_type}, size: {file.size}")
    
    # Read file content to check if it's actually there
    file_content = await file.read()
    content_length = len(file_content) if file_content else 0
    logger.info(f"Actual content length read: {content_length} bytes")
    
    # Reset file position for later use
    await file.seek(0)
    
    if content_length == 0:
        raise FileValidationError(
            message="Empty file received",
            details="The uploaded file contains no data. This may be a browser or network issue.",
            filename=file.filename,
            suggestions=[
                "Try refreshing the page and uploading again",
                "Check if the file is accessible on your system",
                "Try a different browser",
            ],
        )

    # Validate file size before saving
    if file.size and not FileHandler.validate_file_size(file.size):
        file_size_mb = file.size / (1024 * 1024)
        raise FileValidationError(
            message="File too large",
            details=f"File size is {file_size_mb:.1f}MB, maximum allowed is {settings.MAX_UPLOAD_SIZE}MB",
            filename=file.filename,
            file_size_mb=file_size_mb,
            suggestions=[
                "Compress the audio file before uploading",
                "Split the audio into smaller segments",
                "Use a lower quality encoding",
            ],
        )

    # Validate file extension before saving
    if file.filename:
        from pathlib import Path
        ext = Path(file.filename).suffix.lower()
        if ext not in AudioProcessor.SUPPORTED_FORMATS:
            raise FileValidationError(
                message=f"Unsupported file format: {ext}",
                details=f"The file extension {ext} is not supported",
                filename=file.filename,
                expected_formats=AudioProcessor.SUPPORTED_FORMATS_LIST,
                suggestions=[
                    f"Convert to a supported format: {', '.join(AudioProcessor.SUPPORTED_FORMATS_LIST)}",
                    "Use ffmpeg to convert: ffmpeg -i input_file output.wav",
                ],
            )

    # Generate unique filename and save
    unique_filename = FileHandler.generate_unique_filename(file.filename or "audio.wav")
    file_path = await FileHandler.save_upload_file(
        file.file,
        unique_filename,
        settings.UPLOAD_DIR,
    )

    # Validate audio file with detailed error reporting
    try:
        AudioProcessor.validate_audio_file_detailed(file_path)
    except FileValidationError:
        # Cleanup file and re-raise
        AudioProcessor.cleanup_file(file_path)
        raise

    # Create transcription request
    try:
        transcription_request = TranscriptionRequest(
            language=language,
            model_size=ModelSize(model_size),
            enable_diarization=enable_diarization,
            num_speakers=num_speakers,
        )
    except ValueError as e:
        AudioProcessor.cleanup_file(file_path)
        raise FileValidationError(f"Invalid parameters: {str(e)}")

    # Create job
    job_manager = get_job_manager()
    job_id = job_manager.create_job(
        job_type="transcription",
        file_path=file_path,
        parameters=transcription_request.model_dump(),
    )

    # Submit job for processing
    job_manager.submit_transcription_job(
        job_id=job_id,
        file_path=file_path,
        request=transcription_request,
    )

    logger.info(f"Created transcription job {job_id}")

    return JobCreateResponse(
        job_id=job_id,
        status=JobStatus.PENDING,
        message="Transcription job created and queued for processing",
    )


@router.get("/transcribe/job/{job_id}", response_model=JobResponse)
async def get_job_status(job_id: str) -> JobResponse:
    """Get transcription job status and result.

    Args:
        job_id: Job ID to query

    Returns:
        Job status and result if completed

    Raises:
        JobNotFoundError: If job doesn't exist
    """
    logger.debug(f"Fetching status for job {job_id}")
    job_manager = get_job_manager()
    return job_manager.get_job(job_id)


@router.get("/models", response_model=ModelsResponse)
async def get_models() -> ModelsResponse:
    """Get information about available models.

    Returns:
        Available models and supported languages
    """
    service = get_transcription_service()

    return ModelsResponse(
        whisper_models=service.get_available_models(),
        current_model=service.model_size,
        supported_languages=service.get_supported_languages(),
    )


@router.get("/health", response_model=HealthResponse)
async def health_check() -> HealthResponse:
    """Health check for transcription service.

    Returns:
        Service health status
    """
    return HealthResponse(
        status="healthy",
        services={
            "transcription": "ready",
            "diarization": "disabled" if not settings.ENABLE_DIARIZATION else "ready",
            "translation": "disabled" if not settings.ENABLE_TRANSLATION else "ready",
        },
        version="0.1.0",
    )
